File: localizer/localizer.go
package localizer

import (
	"crypto/sha1"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
)

const maxWorkers = 8

var (
	// General URL pattern, excludes newlines
	urlPattern = regexp.MustCompile(`https?://(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}(?:[/?#][^\s'"\` + "`" + `<>()]*)?`)
	// CSS url() pattern
	cssUrlPattern = regexp.MustCompile(`url\((?:['"]?)(https?://.*?)(?:['"]?)\)`)
	// JS string pattern
	jsUrlPattern = regexp.MustCompile(`['"\` + "`" + `](https?://[^\'"` + "`" + `\s\n\r]+)['"\` + "`" + `]`)
	// HTML style tag pattern
	styleTagPattern = regexp.MustCompile(`<style[^>]*>([\s\S]*?)</style>`)
)

// resourceExtensions is a whitelist of acceptable resource file extensions.
var resourceExtensions = map[string]bool{
	".png": true, ".jpg": true, ".jpeg": true, ".gif": true, ".webp": true, ".bmp": true, ".svg": true,
	".mp3": true, ".wav": true, ".ogg": true, ".m4a": true, ".flac": true, ".mid": true,
	".mp4": true, ".webm": true, ".mov": true, ".avi": true,
	".woff": true, ".woff2": true, ".ttf": true, ".otf": true,
	".css": true, ".js": true,
	".json": true, ".txt": true,
}

type downloadTask struct {
	URL       string
	LocalPath string
	Proxy     *url.URL
}

type downloadResult struct {
	Task    downloadTask
	Content []byte
	Err     error
}

// Localizer 处理资源查找、下载和替换的逻辑
type Localizer struct {
	cardData         map[string]interface{}
	outputDir        string
	safeCharName     string
	proxy            *url.URL
	successfulURLMap sync.Map // map[string]string, 从原始 URL 到新的 web 路径
	textContentQueue chan map[string]string
	processedURLs    sync.Map // map[string]bool
	wg               sync.WaitGroup
	stopOnce         sync.Once
	stopChan         chan struct{}
	progressCallback func(message string, level string)
}

// NewLocalizer 创建一个新的 Localizer 实例
func NewLocalizer(cardData map[string]interface{}, outputDir string, proxyStr string, progressCallback func(message string, level string)) (*Localizer, error) {
	var proxyURL *url.URL
	var err error
	if proxyStr != "" {
		proxyURL, err = url.Parse(proxyStr)
		if err != nil {
			return nil, fmt.Errorf("无效的代理 URL: %w", err)
		}
	}

	return &Localizer{
		cardData:         cardData,
		outputDir:        outputDir,
		safeCharName:     filepath.Base(outputDir),
		proxy:            proxyURL,
		textContentQueue: make(chan map[string]string, 100),
		stopChan:         make(chan struct{}),
		progressCallback: progressCallback,
	}, nil
}

// Stop 优雅地停止本地化进程
func (l *Localizer) Stop() {
	l.stopOnce.Do(func() {
		close(l.stopChan)
	})
}

// getResourcePaths 为 URL 生成物理存储路径和 Web 访问路径
func (l *Localizer) getResourcePaths(rawURL, context string) (string, string) {
	parsedURL, err := url.Parse(rawURL)
	if err != nil {
		return "", ""
	}
	parsedPath := parsedURL.Path

	var filename, subDirName string

	// HTML/CSS 上下文逻辑
	if context == "html" || context == "css" {
		hash := sha1.Sum([]byte(rawURL))
		urlHash := hex.EncodeToString(hash[:])[:12]
		fileExt := filepath.Ext(parsedPath)
		if fileExt == "" {
			fileExt = ".dat"
		}
		if strings.Contains(rawURL, "googleapis.com/css") {
			fileExt = ".css"
		}
		filename = urlHash + fileExt
		subDirName = "" // HTML/CSS 资源直接放在角色目录下
	} else { // JS 或其他上下文逻辑
		filename = filepath.Base(parsedPath)
		if filename == "" || filename == "." {
			hash := sha1.Sum([]byte(rawURL))
			fileExt := filepath.Ext(parsedPath)
			if fileExt == "" {
				fileExt = ".dat"
			}
			filename = hex.EncodeToString(hash[:])[:12] + fileExt
		}

		fileExt := strings.ToLower(filepath.Ext(filename))
		switch {
		case fileExt == ".png" || fileExt == ".jpg" || fileExt == ".jpeg" || fileExt == ".webp" || fileExt == ".gif" || fileExt == ".svg" || fileExt == ".bmp":
			subDirName = "images"
		case fileExt == ".mp3" || fileExt == ".wav" || fileExt == ".ogg" || fileExt == ".m4a" || fileExt == ".flac" || fileExt == ".mid" || fileExt == ".mp4" || fileExt == ".webm" || fileExt == ".mov" || fileExt == ".avi":
			subDirName = "media"
		default:
			subDirName = "assets"
		}
	}

	localPath := filepath.Join(l.outputDir, subDirName, filename)
	webPath := "/niko/" + l.safeCharName + "/"
	if subDirName != "" {
		webPath += subDirName + "/"
	}
	webPath += filename

	return localPath, webPath
}

// findAndQueueURLs 在文本内容中查找 URL 并将其加入下载队列
func (l *Localizer) findAndQueueURLs(textContent, context string) []downloadTask {
	rawURLs := make(map[string]bool)

	var patterns []*regexp.Regexp
	switch context {
	case "css":
		patterns = append(patterns, cssUrlPattern, urlPattern)
	case "js":
		patterns = append(patterns, jsUrlPattern)
	case "html":
		// 提取 style 标签内容并加入处理队列
		for _, styleContent := range styleTagPattern.FindAllStringSubmatch(textContent, -1) {
			if len(styleContent) > 1 {
				l.textContentQueue <- map[string]string{"content": styleContent[1], "context": "css"}
			}
		}
		patterns = append(patterns, urlPattern)
	default: // json 或其他
		patterns = append(patterns, urlPattern)
	}

	for _, p := range patterns {
		matches := p.FindAllString(textContent, -1)
		for _, match := range matches {
			// 清理特定模式的匹配结果
			if p == cssUrlPattern {
				match = strings.TrimPrefix(match, "url(")
				match = strings.TrimSuffix(match, ")")
				match = strings.Trim(match, `'"`)
			} else if p == jsUrlPattern {
				match = strings.Trim(match, `'"\`+"`")
			}

			// 处理 JSON 中的多行 URL
			subUrls := regexp.MustCompile(`[\n\r]+|\\n`).Split(match, -1)
			for _, subUrl := range subUrls {
				if strings.TrimSpace(subUrl) != "" {
					rawURLs[subUrl] = true
				}
			}
		}
	}

	var tasks []downloadTask
	for urlStr := range rawURLs {
		cleanedURL := strings.TrimSpace(strings.TrimRight(urlStr, `\`))
		if cleanedURL == "" {
			continue
		}

		if _, loaded := l.processedURLs.LoadOrStore(cleanedURL, true); loaded {
			continue
		}

		parsedURL, err := url.Parse(cleanedURL)
		if err != nil || (parsedURL.Scheme != "http" && parsedURL.Scheme != "https") || parsedURL.Host == "" {
			continue
		}

		fileExt := strings.ToLower(filepath.Ext(parsedURL.Path))
		if !resourceExtensions[fileExt] {
			if !(strings.Contains(cleanedURL, "googleapis.com/css") && (context == "html" || context == "css")) {
				continue
			}
		}

		localPath, _ := l.getResourcePaths(cleanedURL, context)
		if localPath == "" {
			continue
		}

		tasks = append(tasks, downloadTask{
			URL:       cleanedURL,
			LocalPath: localPath,
			Proxy:     l.proxy,
		})
	}
	return tasks
}

// downloadResource 下载单个资源
func downloadResource(task downloadTask) ([]byte, error) {
	if _, err := os.Stat(task.LocalPath); err == nil {
		// 文件已存在，读取它
		content, err := os.ReadFile(task.LocalPath)
		if err == nil {
			return content, nil // 返回已存在的内容
		}
		// 如果读取失败，则继续下载
	}

	if err := os.MkdirAll(filepath.Dir(task.LocalPath), os.ModePerm); err != nil {
		return nil, fmt.Errorf("创建目录失败: %w", err)
	}

	client := &http.Client{}
	if task.Proxy != nil {
		client.Transport = &http.Transport{Proxy: http.ProxyURL(task.Proxy)}
	}

	resp, err := client.Get(task.URL)
	if err != nil {
		return nil, fmt.Errorf("下载失败: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("错误的响应状态: %s", resp.Status)
	}

	content, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("读取响应体失败: %w", err)
	}

	if err := os.WriteFile(task.LocalPath, content, 0644); err != nil {
		return nil, fmt.Errorf("写入文件失败: %w", err)
	}

	return content, nil
}

// Localize 开始本地化进程
func (l *Localizer) Localize() (map[string]interface{}, error) {
	// 初始队列
	cardDataBytes, _ := json.Marshal(l.cardData)
	l.textContentQueue <- map[string]string{"content": string(cardDataBytes), "context": "json"}
	l.wg.Add(1) // 为初始内容添加一个计数

	taskChan := make(chan downloadTask, 100)
	resultChan := make(chan downloadResult, 100)

	// 启动下载工作线程
	for i := 0; i < maxWorkers; i++ {
		go func() {
			for task := range taskChan {
				content, err := downloadResource(task)
				resultChan <- downloadResult{Task: task, Content: content, Err: err}
			}
		}()
	}

	// 主处理循环
	go func() {
		for item := range l.textContentQueue {
			select {
			case <-l.stopChan:
				l.wg.Done() // 如果停止，确保计数器递减
				continue
			default:
				context := item["context"]
				tasks := l.findAndQueueURLs(item["content"], context)
				if len(tasks) > 0 {
					l.progressCallback(fmt.Sprintf("在 %s 上下文中发现 %d 个新URL", context, len(tasks)), "info")
					l.wg.Add(len(tasks))
					for _, task := range tasks {
						taskChan <- task
					}
				}
				l.wg.Done()
			}
		}
	}()

	// 结果处理循环
	go func() {
		for result := range resultChan {
			l.handleDownloadResult(result)
			// 下载任务的 wg.Done() 现在在这里
			l.wg.Done()
		}
	}()

	l.wg.Wait()
	close(taskChan)
	close(resultChan)
	close(l.textContentQueue)

	select {
	case <-l.stopChan:
		l.progressCallback("本地化进程已停止。", "warning")
		return nil, fmt.Errorf("本地化被用户停止")
	default:
	}

	l.progressCallback("所有资源下载完毕，正在替换路径...", "info")
	finalData, ok := l.replaceURLsRecursive(l.cardData).(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("无法将最终数据转换为 map[string]interface{}")
	}
	return finalData, nil
}

func (l *Localizer) handleDownloadResult(result downloadResult) {
	if result.Err != nil {
		l.progressCallback(fmt.Sprintf("[失败] %s - %v", result.Task.URL, result.Err), "failure")
		return
	}

	l.progressCallback(fmt.Sprintf("[成功] %s", result.Task.URL), "success")

	_, webPath := l.getResourcePaths(result.Task.URL, "") // 上下文对于 web 路径生成不重要
	if webPath != "" {
		l.successfulURLMap.Store(result.Task.URL, webPath)
	}

	// 如果下载的文件是文本，则将其添加到队列中进行递归处理
	ext := strings.ToLower(filepath.Ext(result.Task.LocalPath))
	if ext == ".css" || ext == ".js" || ext == ".html" || ext == ".htm" {
		l.wg.Add(1)
		l.textContentQueue <- map[string]string{"content": string(result.Content), "context": strings.TrimPrefix(ext, ".")}
	}
}

// replaceURLsRecursive 递归地替换数据结构中的 URL
func (l *Localizer) replaceURLsRecursive(data interface{}) interface{} {
	switch v := data.(type) {
	case map[string]interface{}:
		newMap := make(map[string]interface{})
		for key, val := range v {
			newMap[key] = l.replaceURLsRecursive(val)
		}
		return newMap
	case []interface{}:
		newSlice := make([]interface{}, len(v))
		for i, val := range v {
			newSlice[i] = l.replaceURLsRecursive(val)
		}
		return newSlice
	case string:
		// 这是一个简单的替换。对于单个字符串可能包含多个 URL 或其周围有文本的更复杂场景，
		// 需要使用更复杂的、基于正则表达式的替换方法。
		l.successfulURLMap.Range(func(key, value interface{}) bool {
			v = strings.ReplaceAll(v, key.(string), value.(string))
			return true
		})
		return v
	default:
		return data
	}
}

File: 角色卡本地化/localizer.py
# -*- coding: utf-8 -*-

"""
角色卡本地化工具
localizer.py - 资源查找、下载和替换的核心逻辑 (v3.0 - 遵照原始逻辑修复版)
"""

import re
import requests
import threading
import hashlib
from pathlib import Path
from urllib.parse import urlparse, quote
from concurrent.futures import ThreadPoolExecutor, as_completed
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

import importlib.util

try:
    # 检查 socks 模块是否可用
    if importlib.util.find_spec("socks") and importlib.util.find_spec("socket"):
        import socks

        # 进一步检查是否是 PySocks，因为它提供了 SOCKS5 常量
        if hasattr(socks, "SOCKS5"):
            SOCKS_SUPPORTED = True
        else:
            SOCKS_SUPPORTED = False
    else:
        SOCKS_SUPPORTED = False
except ImportError:
    SOCKS_SUPPORTED = False

MAX_WORKERS = 8

# 定义可接受的资源文件扩展名白名单
RESOURCE_EXTENSIONS = {
    # Images
    ".png",
    ".jpg",
    ".jpeg",
    ".gif",
    ".webp",
    ".bmp",
    ".svg",
    # Audio
    ".mp3",
    ".wav",
    ".ogg",
    ".m4a",
    ".flac",
    ".mid",
    # Video
    ".mp4",
    ".webm",
    ".mov",
    ".avi",
    # Fonts
    ".woff",
    ".woff2",
    ".ttf",
    ".otf",
    # Styles & Scripts
    ".css",
    ".js",
    # Data
    ".json",
    ".txt",
}


def create_session(proxies=None):
    session = requests.Session()
    retries = Retry(total=2, backoff_factor=0.5, status_forcelist=[500, 502, 503, 504])
    adapter = HTTPAdapter(max_retries=retries)
    session.mount("http://", adapter)
    session.mount("https", adapter)
    if proxies:
        session.proxies = proxies
    session.headers.update({"User-Agent": "Mozilla/5.0"})
    return session


def download_resource(task_info):
    url, local_path, proxies, force_proxy = task_info.values()

    # 简单地对URL中的空格等进行编码，大多数现代库能处理好其他字符
    encoded_url = quote(url, safe="/:@?=&")

    local_path.parent.mkdir(parents=True, exist_ok=True)
    if local_path.exists():
        try:
            content = local_path.read_bytes()
            return {
                "status": "skipped",
                "method": "文件已存在",
                "url": url,
                "local_path": local_path,
                "content": content,
            }
        except Exception:
            # 如果读取已存在文件失败，当作不存在处理
            pass

    session_args = {"timeout": 15}
    if not force_proxy:
        try:
            response = create_session().get(encoded_url, **session_args)
            response.raise_for_status()
            local_path.write_bytes(response.content)
            return {
                "status": "success",
                "method": "直连",
                "url": url,
                "local_path": local_path,
                "content": response.content,
            }
        except requests.RequestException:
            pass

    if proxies:
        try:
            method = "强制代理" if force_proxy else "代理回退"
            session_args["timeout"] = 45
            response = create_session(proxies=proxies).get(encoded_url, **session_args)
            response.raise_for_status()
            local_path.write_bytes(response.content)
            return {
                "status": "success",
                "method": method,
                "url": url,
                "local_path": local_path,
                "content": response.content,
            }
        except requests.RequestException as e:
            error_message = str(e).splitlines()[0] if str(e) else "未知错误"
            return {
                "status": "failure",
                "url": url,
                "error": f"代理失败: {error_message}",
            }

    return {"status": "failure", "url": url, "error": "直连失败且未提供代理"}


class Localizer:
    def __init__(self, card_data, output_dir, proxy=None, force_proxy_domains=None):
        self.card_data = card_data
        self.output_dir = Path(output_dir)
        # 从 output_dir 推断角色名，用于构建 web 路径
        self.safe_char_name = self.output_dir.name
        if (
            proxy
            and proxy.startswith(("socks5://", "socks5h://"))
            and not SOCKS_SUPPORTED
        ):
            raise ImportError(
                "检测到SOCKS代理，但未安装 PySocks 库。请运行 'pip install PySocks'。"
            )
        self.proxies = {"http": proxy, "https": proxy} if proxy else None
        self.force_proxy_domains = (
            force_proxy_domains if force_proxy_domains is not None else []
        )

        # 修正：在通用模式中明确排除换行符，防止跨行匹配
        self.url_pattern = re.compile(r'https?://[^\s\'"`<>(),\n\r]+')
        self.css_url_pattern = re.compile(r'url\((?:[\'"]?)(https?://.*?)(?:[\'"]?)\)')
        # 修正：在JS模式中也明确排除换行符
        self.js_url_pattern = re.compile(r'[\'"`](https?://[^\'"`\s\n\r]+)[\'"`]')
        # 用于从HTML中提取style标签内容的模式
        self.style_tag_pattern = re.compile(r"<style[^>]*>([\s\S]*?)</style>")

        self.successful_url_map = {}  # { "http://...": "/niko/char_name/file.css" }
        self.text_content_queue = []  # 待处理的文本内容
        self.processed_urls = set()  # 已处理过的URL，防止重复下载
        self._stop_event = threading.Event()  # 用于停止本地化进程的事件

    def get_resource_paths(self, url, context):
        """
        遵照参考脚本逻辑，为URL生成物理存储路径和Web访问路径。
        """
        try:
            parsed_url = urlparse(url)
            parsed_path = Path(parsed_url.path)

            # HTML/CSS 上下文逻辑 (来自 html资源本地化.py)
            if context in ["html", "css"]:
                url_hash = hashlib.sha1(url.encode()).hexdigest()[:12]
                file_ext = parsed_path.suffix or ".dat"
                if "googleapis.com/css" in url:
                    file_ext = ".css"

                filename = f"{url_hash}{file_ext}"
                # HTML资源直接放在角色目录下
                local_path = self.output_dir / filename
                # Web路径是绝对路径
                web_path = f"/niko/{self.safe_char_name}/{filename}"
                return local_path, web_path

            # JS 或其他上下文逻辑 (来自 js代码内置资源本地化.py)
            else:
                filename = parsed_path.name
                if not filename:  # 备用方案
                    filename = hashlib.sha1(url.encode()).hexdigest()[:12] + (
                        parsed_path.suffix or ".dat"
                    )

                # 根据文件类型确定子目录
                file_ext = parsed_path.suffix.lower()
                if file_ext in [
                    ".png",
                    ".jpg",
                    ".jpeg",
                    ".webp",
                    ".gif",
                    ".svg",
                    ".bmp",
                ]:
                    sub_dir_name = "images"
                elif file_ext in [
                    ".mp3",
                    ".wav",
                    ".ogg",
                    ".m4a",
                    ".flac",
                    ".mid",
                    ".mp4",
                    ".webm",
                    ".mov",
                    ".avi",
                ]:
                    sub_dir_name = "media"
                else:  # 字体, css, js, json等
                    sub_dir_name = "assets"

                local_path = self.output_dir / sub_dir_name / filename
                web_path = f"/niko/{self.safe_char_name}/{sub_dir_name}/{filename}"
                return local_path, web_path

        except Exception:
            return None, None

    def find_and_queue_urls(self, text_content, context):
        """根据上下文查找URL并加入下载队列"""
        raw_urls = set()
        # 1. 根据不同上下文使用不同策略
        if context == "css":
            raw_urls.update(
                m.group(1) for m in self.css_url_pattern.finditer(text_content)
            )
            raw_urls.update(
                self.url_pattern.findall(text_content)
            )  # CSS中也可能直接写URL
        elif context == "js":
            raw_urls.update(
                m.group(1) for m in self.js_url_pattern.finditer(text_content)
            )
        elif context == "html":
            # 提取HTML中的style标签内容，并加入处理队列
            for style_content in self.style_tag_pattern.findall(text_content):
                self.text_content_queue.append(
                    {"content": style_content, "context": "css"}
                )
            raw_urls.update(self.url_pattern.findall(text_content))
        else:  # json 或其他
            # 修正：处理JSON字符串化后，一个字段内含多个URL（以\n或\\n分隔）的情况
            potential_url_blobs = self.url_pattern.findall(text_content)
            for blob in potential_url_blobs:
                # 对每个文本块按换行符（包括转义的）进行分割
                urls_in_blob = re.split(r"[\n\r]+|\\n", blob)
                for url in urls_in_blob:
                    if url.strip():
                        raw_urls.add(url.strip())

        # 3. 清理URL (由于正则已修正，不再需要复杂的分割逻辑)
        tasks = []
        for url in raw_urls:
            cleaned_url = url.strip().rstrip("\\")
            if not cleaned_url or cleaned_url in self.processed_urls:
                continue

            # 4. 严格的URL过滤
            try:
                parsed_url = urlparse(url)
                # 必须是 http/https 协议，并且有域名
                if parsed_url.scheme not in ["http", "https"] or not parsed_url.netloc:
                    continue
                # 必须有路径，并且路径的扩展名在白名单内
                file_ext = Path(parsed_url.path).suffix.lower()
                if not file_ext or file_ext not in RESOURCE_EXTENSIONS:
                    # 特例：处理 Google Fonts 这种没有扩展名的CSS链接
                    if not ("googleapis.com/css" in url and context in ["html", "css"]):
                        continue
            except Exception:
                continue  # 解析失败的URL直接跳过

            self.processed_urls.add(cleaned_url)

            # 使用新的统一路径生成函数
            local_path, _ = self.get_resource_paths(cleaned_url, context)
            if not local_path:
                continue

            force_proxy = any(
                domain in cleaned_url for domain in self.force_proxy_domains
            )
            tasks.append(
                {
                    "url": cleaned_url,
                    "local_path": local_path,
                    "proxies": self.proxies,
                    "force_proxy": force_proxy,
                }
            )
        return tasks

    def localize_with_progress(self, progress_signal):
        # 1. 初始化队列
        self.text_content_queue.append(
            {"content": str(self.card_data), "context": "json"}
        )

        # 2. 循环处理队列，直到没有新的文本内容需要处理
        while self.text_content_queue and not self._stop_event.is_set():
            item = self.text_content_queue.pop(0)
            text_content, context = item["content"], item["context"]

            tasks = self.find_and_queue_urls(text_content, context)
            if not tasks:
                continue

            progress_signal.emit(
                f"在 {context} 上下文中发现 {len(tasks)} 个新URL，开始下载...", "info"
            )

            with ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
                futures = {executor.submit(download_resource, task) for task in tasks}
                for future in as_completed(futures):
                    if self._stop_event.is_set():
                        executor.shutdown(
                            wait=False, cancel_futures=True
                        )  # 尝试取消未完成的任务
                        break  # 退出循环
                    result = future.result()
                    status, url = result["status"], result["url"]

                    method = result.get("method", "失败")
                    msg = f"[{method}] {url}"
                    if status == "failure":
                        msg += f" - {result['error']}"
                    progress_signal.emit(msg, status)

                    if status in ["success", "skipped"]:
                        # 使用新的统一路径生成函数获取Web路径
                        _, web_path = self.get_resource_paths(url, context)
                        if web_path:
                            self.successful_url_map[url] = web_path

                        # 如果下载的是文本文件，加入队列进行递归处理
                        content = result.get("content", b"")
                        if content and result["local_path"].suffix.lower() in [
                            ".css",
                            ".js",
                            ".html",
                            ".htm",
                        ]:
                            try:
                                new_context = (
                                    result["local_path"].suffix.lower().strip(".")
                                )
                                self.text_content_queue.append(
                                    {
                                        "content": content.decode("utf-8"),
                                        "context": new_context,
                                    }
                                )
                            except (UnicodeDecodeError, AttributeError):
                                pass

        # 3. 所有资源下载完毕后，进行最终的全局替换
        progress_signal.emit("所有资源下载完毕，正在替换路径...", "info")
        if self._stop_event.is_set():
            progress_signal.emit("本地化进程已终止，跳过路径替换。", "warning")
            return None

        final_data = self._replace_urls_recursive(
            self.card_data, self.successful_url_map
        )

        if self._stop_event.is_set():
            progress_signal.emit("路径替换过程中止。", "warning")
            return None
        else:
            progress_signal.emit("所有阶段处理完毕。", "success")
            return final_data

    def _replace_urls_recursive(self, data, url_map):
        if isinstance(data, dict):
            return {
                k: self._replace_urls_recursive(v, url_map) for k, v in data.items()
            }
        elif isinstance(data, list):
            return [self._replace_urls_recursive(i, url_map) for i in data]
        elif isinstance(data, str):
            for url, local_path in url_map.items():
                data = data.replace(url, local_path)
            return data

        return data

    def stop(self):
        self._stop_event.set()
